File: DiscordBridge-node/src/protocol.ts
// Wire protocol mirror of DiscordAPI/Protocol.cs. Keep both sides in sync.
// PROTOCOL_VERSION here must match ProtocolConstants.Version on the C# side.
//
// Two frame shapes share the outer 4-byte LE length prefix; the first byte of
// the payload tells them apart:
//
//   0x7B ('{')  → UTF-8 JSON, dispatched by op string. Used for everything
//                 except SpeakPcm.
//   0x01        → Binary SpeakPcm, plugin → bridge only:
//                   [0x01]
//                   [reqId u32 LE]
//                   [sampleRate u32 LE]
//                   [bits u8]
//                   [channels u8]
//                   [raw PCM bytes...]   // remainder of payload
//                 Header is 11 bytes. Response is JSON `SpeakResult` with the
//                 matching reqId.
//
// SpeakFile is a normal JSON op carrying a file path; the bridge opens and
// streams the file itself (must be 48 kHz / 16-bit / stereo PCM WAV).

export const PROTOCOL_VERSION = 1 as const;
export const MAX_FRAME_BYTES = 64 * 1024 * 1024;

export const FRAME_JSON_MARKER = 0x7B; // '{'
export const FRAME_BINARY_SPEAK_PCM = 0x01;
export const BINARY_SPEAK_PCM_HEADER_BYTES = 11;

export const Op = {
    Hello: 'Hello', HelloResult: 'HelloResult',
    Init: 'Init', InitResult: 'InitResult',
    Deinit: 'Deinit', DeinitResult: 'DeinitResult',
    IsConnected: 'IsConnected', IsConnectedResult: 'IsConnectedResult',
    GetServers: 'GetServers', GetServersResult: 'GetServersResult',
    GetChannels: 'GetChannels', GetChannelsResult: 'GetChannelsResult',
    SetGame: 'SetGame', SetGameResult: 'SetGameResult',
    JoinChannel: 'JoinChannel', JoinChannelResult: 'JoinChannelResult',
    LeaveChannel: 'LeaveChannel', LeaveChannelResult: 'LeaveChannelResult',
    SpeakPcm: 'SpeakPcm',
    SpeakFile: 'SpeakFile',
    SpeakResult: 'SpeakResult',
    Shutdown: 'Shutdown',
    BotReady: 'BotReady', Log: 'Log', Disconnected: 'Disconnected',
} as const;

export type OpName = typeof Op[keyof typeof Op];

export type LogLevel = 'Info' | 'Warn' | 'Error';

export type ReqId = number | null;

export interface BaseRequest { op: OpName; reqId: ReqId }

export interface HelloRequest        extends BaseRequest { op: 'Hello'; protocolVersion: number }
export interface InitRequest         extends BaseRequest { op: 'Init'; token: string; status: string }
export interface DeinitRequest       extends BaseRequest { op: 'Deinit' }
export interface IsConnectedRequest  extends BaseRequest { op: 'IsConnected' }
export interface GetServersRequest   extends BaseRequest { op: 'GetServers' }
export interface GetChannelsRequest  extends BaseRequest { op: 'GetChannels'; server: string }
export interface SetGameRequest      extends BaseRequest { op: 'SetGame'; text: string }
export interface JoinChannelRequest  extends BaseRequest { op: 'JoinChannel'; server: string; channel: string }
export interface LeaveChannelRequest extends BaseRequest { op: 'LeaveChannel' }
export interface SpeakFileRequest    extends BaseRequest { op: 'SpeakFile'; path: string }
export interface ShutdownRequest     extends BaseRequest { op: 'Shutdown' }

export type Request =
    | HelloRequest | InitRequest | DeinitRequest | IsConnectedRequest
    | GetServersRequest | GetChannelsRequest | SetGameRequest
    | JoinChannelRequest | LeaveChannelRequest | SpeakFileRequest | ShutdownRequest;

export interface HelloResponse        { op: 'HelloResult';        reqId: ReqId; ok: boolean; bridgeVersion: string; error: string }
export interface InitResponse         { op: 'InitResult';         reqId: ReqId; ok: boolean; error: string }
export interface DeinitResponse       { op: 'DeinitResult';       reqId: ReqId; ok: true;    error: '' }
export interface IsConnectedResponse  { op: 'IsConnectedResult';  reqId: ReqId; connected: boolean }
export interface GetServersResponse   { op: 'GetServersResult';   reqId: ReqId; servers: string[] }
export interface GetChannelsResponse  { op: 'GetChannelsResult';  reqId: ReqId; channels: string[] }
export interface SetGameResponse      { op: 'SetGameResult';      reqId: ReqId; ok: true;    error: '' }
export interface JoinChannelResponse  { op: 'JoinChannelResult';  reqId: ReqId; ok: boolean; error: string }
export interface LeaveChannelResponse { op: 'LeaveChannelResult'; reqId: ReqId; ok: true;    error: '' }
export interface SpeakResponse        { op: 'SpeakResult';        reqId: ReqId; ok: boolean; error: string }

// Generic shape used for the catch-all error response in pipe-server. Matches
// C# OkResponse: any *Result op with reqId/ok/error fields parses as this.
export interface ErrorResponse { op: OpName; reqId: ReqId; ok: false; error: string }

export type Response =
    | HelloResponse | InitResponse | DeinitResponse | IsConnectedResponse
    | GetServersResponse | GetChannelsResponse | SetGameResponse
    | JoinChannelResponse | LeaveChannelResponse | SpeakResponse;

export interface BotReadyNotification     { op: 'BotReady' }
export interface LogNotification          { op: 'Log'; level: LogLevel; message: string }
export interface DisconnectedNotification { op: 'Disconnected'; reason: string }

export type Notification = BotReadyNotification | LogNotification | DisconnectedNotification;

export type OutboundFrame = Response | ErrorResponse | Notification;

File: DiscordAPI/Protocol.cs
using System.Text.Json.Serialization;

namespace DiscordBridge.Protocol {

    public static class ProtocolConstants {
        public const int Version = 2;
    }

    public static class Op {
        public const string Hello = "Hello";
        public const string HelloResult = "HelloResult";
        public const string Init = "Init";
        public const string InitResult = "InitResult";
        public const string Deinit = "Deinit";
        public const string DeinitResult = "DeinitResult";
        public const string IsConnected = "IsConnected";
        public const string IsConnectedResult = "IsConnectedResult";
        public const string GetServers = "GetServers";
        public const string GetServersResult = "GetServersResult";
        public const string GetChannels = "GetChannels";
        public const string GetChannelsResult = "GetChannelsResult";
        public const string SetGame = "SetGame";
        public const string SetGameResult = "SetGameResult";
        public const string JoinChannel = "JoinChannel";
        public const string JoinChannelResult = "JoinChannelResult";
        public const string LeaveChannel = "LeaveChannel";
        public const string LeaveChannelResult = "LeaveChannelResult";
        public const string SpeakPcm = "SpeakPcm";
        public const string SpeakFile = "SpeakFile";
        public const string SpeakResult = "SpeakResult";
        public const string Shutdown = "Shutdown";

        public const string BotReady = "BotReady";
        public const string Log = "Log";
        public const string Disconnected = "Disconnected";
    }

    // Marker interface for request DTOs. Lets PipeClient.SendAsync set ReqId
    // without reflection.
    public interface IBridgeRequest {
        int? ReqId { get; set; }
    }

    public class HelloRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.Hello;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("protocolVersion")] public int ProtocolVersion { get; set; }
    }

    public class DeinitRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.Deinit;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
    }

    public class IsConnectedRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.IsConnected;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
    }

    public class GetServersRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.GetServers;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
    }

    public class LeaveChannelRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.LeaveChannel;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
    }

    public class ShutdownRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.Shutdown;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
    }

    public class HelloResponse {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.HelloResult;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("ok")] public bool Ok { get; set; }
        [JsonPropertyName("bridgeVersion")] public string BridgeVersion { get; set; } = "";
        [JsonPropertyName("error")] public string Error { get; set; } = "";
    }

    public class InitRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.Init;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("token")] public string Token { get; set; } = "";
        [JsonPropertyName("status")] public string Status { get; set; } = "";
    }

    public class GetChannelsRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.GetChannels;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("server")] public string Server { get; set; } = "";
    }

    public class GetChannelsResponse {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.GetChannelsResult;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("channels")] public string[] Channels { get; set; } = new string[0];
    }

    public class GetServersResponse {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.GetServersResult;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("servers")] public string[] Servers { get; set; } = new string[0];
    }

    public class IsConnectedResponse {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.IsConnectedResult;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("connected")] public bool Connected { get; set; }
    }

    public class SetGameRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.SetGame;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("text")] public string Text { get; set; } = "";
    }

    public class JoinChannelRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.JoinChannel;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("server")] public string Server { get; set; } = "";
        [JsonPropertyName("channel")] public string Channel { get; set; } = "";
    }

    public class JoinChannelResponse {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.JoinChannelResult;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("ok")] public bool Ok { get; set; }
        [JsonPropertyName("error")] public string Error { get; set; } = "";
    }

    // SpeakPcm is sent as a length-prefixed BINARY frame, not JSON.
    // See PipeClient.SendSpeakPcmAsync / pipe-server.ts _handleBinarySpeakPcm.
    // Layout (after the outer 4-byte LE length): [0x01][reqId u32 LE][sampleRate u32 LE][bits u8][channels u8][flags u8][raw PCM...]
    // flags bit0 = apply a random sound effect. Response stays JSON: { op:"SpeakResult", reqId, ok, error }.

    public class SpeakFileRequest : IBridgeRequest {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.SpeakFile;
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("path")] public string Path { get; set; } = "";
        // Mirrors the binary SpeakPcm flags bit0: apply a random sound effect to this trigger.
        [JsonPropertyName("randomEffect")] public bool RandomEffect { get; set; }
    }

    public class OkResponse {
        [JsonPropertyName("op")] public string Op { get; set; } = "";
        [JsonPropertyName("reqId")] public int? ReqId { get; set; }
        [JsonPropertyName("ok")] public bool Ok { get; set; }
        [JsonPropertyName("error")] public string Error { get; set; } = "";
    }

    public class LogNotification {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.Log;
        [JsonPropertyName("message")] public string Message { get; set; } = "";
        [JsonPropertyName("level")] public string Level { get; set; } = "Info";
    }

    public class BotReadyNotification {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.BotReady;
    }

    public class DisconnectedNotification {
        [JsonPropertyName("op")] public string Op { get; set; } = Protocol.Op.Disconnected;
        [JsonPropertyName("reason")] public string Reason { get; set; } = "";
    }
}
